File: packages/react-native/ios/VoiceRegistry.swift
import Foundation

/// The native voice sources a React Native app can create (src/voice.ts), kept by
/// id so a view can bind one with the `voiceSourceId` prop and several views can
/// share it. The app owns the lifecycle: nothing here connects on its own.
///
/// Vendors are opt-in per build. The Gemini and ElevenLabs glue is Foundation-only
/// and always present; LiveKit and OpenAI Realtime come from the `SinuaCore/LiveKit`
/// and `SinuaCore/OpenAI` subspecs, which also define SINUA_LIVEKIT / SINUA_OPENAI.
/// Without the subspec, `create` fails with what to add -- nothing is linked silently.
///
/// Credentials live only in the source they are handed to: never stored, never logged.
@MainActor
final class VoiceRegistry {
    static let shared = VoiceRegistry()

    enum RegistryError: LocalizedError {
        case unknownVendor(String)
        case vendorNotInstalled(vendor: String, subspec: String)
        case missingField(vendor: String, field: String)
        case unknownSource(String)

        var errorDescription: String? {
            switch self {
            case .unknownVendor(let v): return "unknown voice vendor \"\(v)\""
            case .vendorNotInstalled(let v, let s):
                return "the \(v) voice source isn't in this build: add `pod 'SinuaCore/\(s)'` (and LiveKit's podspecs source) to your Podfile, then pod install"
            case .missingField(let v, let f): return "\(v) needs \(f)"
            case .unknownSource(let id): return "no voice source \(id) (already released?)"
            }
        }
    }

    private var sources: [String: VoiceSource] = [:]
    /// Waiting `getCredential` round trips (OpenAI): requestId -> continuation.
    private var pending: [String: CheckedContinuation<String, Error>] = [:]

    /// Emitted state / error / interrupt / credentialRequest; the module forwards them to JS.
    var onEvent: ((_ id: String, _ event: String, _ payload: [String: Any]) -> Void)?

    func source(id: String) -> VoiceSource? { sources[id] }

    func create(config: [String: Any]) throws {
        guard let id = config["id"] as? String, let vendor = config["vendor"] as? String else {
            throw RegistryError.missingField(vendor: "voice", field: "id and vendor")
        }
        let source = try make(vendor: vendor, id: id, config: config)
        source.onStateChange { [weak self] state in
            self?.onEvent?(id, "state", ["state": state.rawValue])
        }
        source.onInterrupt { [weak self] in self?.onEvent?(id, "interrupt", [:]) }
        sources[id] = source
    }

    func connect(id: String) async throws {
        guard let source = sources[id] else { throw RegistryError.unknownSource(id) }
        try await source.connect()
    }

    func disconnect(id: String) {
        sources[id]?.disconnect()
    }

    func release(id: String) {
        sources[id]?.disconnect()
        sources[id] = nil
    }

    // MARK: - getCredential round trip (OpenAI: an `ek_` is single-use, so every reconnect asks again)

    func requestCredential(id: String) async throws -> String {
        let requestId = UUID().uuidString
        return try await withCheckedThrowingContinuation { cont in
            pending[requestId] = cont
            onEvent?(id, "credentialRequest", ["requestId": requestId])
        }
    }

    func provideCredential(requestId: String, credential: String?, error: String?) {
        guard let cont = pending.removeValue(forKey: requestId) else { return }
        if let credential, !credential.isEmpty {
            cont.resume(returning: credential)
        } else {
            cont.resume(throwing: RegistryError.missingField(vendor: "openai", field: error ?? "a credential"))
        }
    }

    // MARK: - Vendors

    private func make(vendor: String, id: String, config: [String: Any]) throws -> VoiceSource {
        let string = { (key: String) -> String? in (config[key] as? String).flatMap { $0.isEmpty ? nil : $0 } }
        let allowInsecureApiKey = config["allowInsecureApiKey"] as? Bool ?? false
        switch vendor {
        case "test": return TestToneVoiceSource()
        case "mic": return LocalMicVoiceSource()
        case "gemini":
            guard let credential = string("credential") else { throw RegistryError.missingField(vendor: "gemini", field: "a credential") }
            return GeminiLiveVoiceSource(credential: credential,
                                         model: string("model") ?? GeminiLiveSession.defaultModel,
                                         instructions: string("instructions"),
                                         allowInsecureApiKey: allowInsecureApiKey,
                                         endpoint: string("endpoint").flatMap(URL.init(string:)).map { GeminiLiveSession.Endpoint(url: $0, headers: [:]) })
        case "elevenlabs":
            guard let credential = string("credential") else { throw RegistryError.missingField(vendor: "elevenlabs", field: "an agent id or signed URL") }
            return ElevenLabsVoiceSource(credential: credential, endpoint: string("endpoint").flatMap { URL(string: $0) })
        case "livekit":
            #if SINUA_LIVEKIT
            guard let url = string("url"), let token = string("token") else {
                throw RegistryError.missingField(vendor: "livekit", field: "a url and token")
            }
            return LiveKitVoiceSource(url: url, token: token, publishMicrophone: config["publishMicrophone"] as? Bool ?? true)
            #else
            throw RegistryError.vendorNotInstalled(vendor: "LiveKit", subspec: "LiveKit")
            #endif
        case "openai":
            #if SINUA_OPENAI
            if config["hasCredentialProvider"] as? Bool == true {
                return OpenAIRealtimeVoiceSource(credentialProvider: { [weak self] in
                    guard let self else { throw RegistryError.unknownSource(id) }
                    return try await self.requestCredential(id: id)
                })
            }
            guard let credential = string("credential") else {
                throw RegistryError.missingField(vendor: "openai", field: "a credential or getCredential")
            }
            return OpenAIRealtimeVoiceSource(credential: credential,
                                             model: string("model") ?? OpenAIRealtimeSignaling.defaultModel,
                                             voice: string("voice") ?? OpenAIRealtimeSignaling.defaultVoice,
                                             instructions: string("instructions"),
                                             allowInsecureApiKey: allowInsecureApiKey)
            #else
            throw RegistryError.vendorNotInstalled(vendor: "OpenAI Realtime", subspec: "OpenAI")
            #endif
        default:
            throw RegistryError.unknownVendor(vendor)
        }
    }
}

File: packages/react-native/android/src/main/java/dev/sinua/reactnative/VoiceRegistry.kt
package dev.sinua.reactnative

import android.content.Context
import dev.sinua.voice.LocalMicVoiceSource
import dev.sinua.voice.TestToneVoiceSource
import dev.sinua.voice.VoiceSource
import com.facebook.react.bridge.ReadableMap
import kotlinx.coroutines.CompletableDeferred

/**
 * The native voice sources a React Native app can create (src/voice.ts), kept by
 * id so a view can bind one with the `voiceSourceId` prop and several views can
 * share it. The app owns the lifecycle: nothing here connects on its own.
 *
 * Vendors are opt-in per build: `sinua.voiceVendors` in the app's
 * gradle.properties selects which vendor sources and dependencies are compiled in
 * (default `gemini,elevenlabs`, which need nothing beyond OkHttp). A vendor that
 * isn't in the build has no factory class, and `create` says how to add it --
 * nothing is linked silently.
 *
 * Credentials live only in the source they are handed to: never stored, never logged.
 */
object VoiceRegistry {
    /** Implemented once per vendor, in that vendor's source set (src/<vendor>/kotlin). */
    interface VendorFactory {
        /**
         * [errors] reports failures that happen *after* connect() returned (a dropped
         * socket, a vendor error frame); the JS handle surfaces them as `error` events.
         */
        fun create(context: Context, config: ReadableMap, credentials: CredentialProvider, errors: (String) -> Unit): VoiceSource
    }

    /** OpenAI's `getCredential`: asked again for every (re)connect, since an `ek_` is single-use. */
    fun interface CredentialProvider {
        suspend fun fetch(): String
    }

    class VoiceError(message: String) : Exception(message)

    private val sources = mutableMapOf<String, VoiceSource>()
    private val pending = mutableMapOf<String, CompletableDeferred<String>>()

    /** Emitted state / error / interrupt / credentialRequest; the module forwards them to JS. */
    var onEvent: ((id: String, event: String, payload: Map<String, Any>) -> Unit)? = null

    fun source(id: String): VoiceSource? = synchronized(sources) { sources[id] }

    fun create(context: Context, config: ReadableMap) {
        val id = config.getString("id") ?: throw VoiceError("voice source needs an id")
        val vendor = config.getString("vendor") ?: throw VoiceError("voice source needs a vendor")
        val source = make(context, vendor, id, config)
        source.onStateChange { state -> onEvent?.invoke(id, "state", mapOf("state" to state.wire)) }
        source.onInterrupt { onEvent?.invoke(id, "interrupt", emptyMap()) }
        synchronized(sources) { sources[id] = source }
    }

    fun connect(id: String) {
        (source(id) ?: throw VoiceError("no voice source $id (already released?)")).connect()
    }

    fun disconnect(id: String) {
        source(id)?.disconnect()
    }

    fun release(id: String) {
        synchronized(sources) { sources.remove(id) }?.disconnect()
    }

    // MARK: getCredential round trip

    suspend fun requestCredential(id: String): String {
        val requestId = java.util.UUID.randomUUID().toString()
        val waiter = CompletableDeferred<String>()
        synchronized(pending) { pending[requestId] = waiter }
        onEvent?.invoke(id, "credentialRequest", mapOf("requestId" to requestId))
        return waiter.await()
    }

    fun provideCredential(requestId: String, credential: String?, error: String?) {
        val waiter = synchronized(pending) { pending.remove(requestId) } ?: return
        if (!credential.isNullOrEmpty()) waiter.complete(credential)
        else waiter.completeExceptionally(VoiceError(error ?: "getCredential returned no credential"))
    }

    // MARK: vendors

    private fun make(context: Context, vendor: String, id: String, config: ReadableMap): VoiceSource = when (vendor) {
        "test" -> TestToneVoiceSource()
        "mic" -> LocalMicVoiceSource()
        else -> factory(vendor).create(context, config, CredentialProvider { requestCredential(id) }) { message ->
            onEvent?.invoke(id, "error", mapOf("message" to message))
        }
    }

    private fun factory(vendor: String): VendorFactory {
        val className = "dev.sinua.reactnative.vendors." + vendor.replaceFirstChar { it.uppercase() } + "Factory"
        val cls = try {
            Class.forName(className)
        } catch (_: ClassNotFoundException) {
            throw VoiceError(
                "the $vendor voice source isn't in this build: add it to `sinua.voiceVendors` " +
                    "in your app's gradle.properties (e.g. sinua.voiceVendors=gemini,elevenlabs,$vendor) and rebuild",
            )
        }
        return cls.getDeclaredConstructor().newInstance() as VendorFactory
    }
}
